internal/service/dataexchange/event_action.go

```diff
@@ -0,0 +1,416 @@
+	"github.com/hashicorp/terraform-provider-aws/internal/errs"
+	"github.com/hashicorp/terraform-provider-aws/internal/framework"
+	fwtypes "github.com/hashicorp/terraform-provider-aws/internal/framework/types"
+	"github.com/hashicorp/terraform-provider-aws/internal/framework/validators"
+	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
+	"github.com/hashicorp/terraform-provider-aws/names"
+)
+
+// @FrameworkResource("aws_dataexchange_event_action", name="Event Action")
+func ResourceEventAction(_ context.Context) (resource.ResourceWithConfigure, error) {
+	r := &resourceEventAction{}
+
+	r.SetDefaultCreateTimeout(30 * time.Minute)
+	r.SetDefaultUpdateTimeout(30 * time.Minute)
+	r.SetDefaultDeleteTimeout(30 * time.Minute)
+
+	return r, nil
+}
+
+const (
+	ResNameEventAction = "Event Action"
+)
+
+type resourceEventAction struct {
+	framework.ResourceWithConfigure
+	framework.WithTimeouts
+}
+
+func (r *resourceEventAction) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = "aws_dataexchange_event_action"
+}
+
+func (r *resourceEventAction) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			names.AttrARN: framework.ARNAttributeComputedOnly(),
+			names.AttrID:  framework.IDAttribute(),
+		},
+		Blocks: map[string]schema.Block{
+			"action_export_revision_to_s3": schema.SingleNestedBlock{
+				CustomType: fwtypes.NewObjectTypeOf[actionExportRevisionToS3Model](ctx),
+				Attributes: map[string]schema.Attribute{
+					"bucket": schema.StringAttribute{
+						Required: true,
+					},
+					"key_pattern": schema.StringAttribute{
+						Optional: true,
+						Computed: true,
+						Default:  stringdefault.StaticString("${Revision.CreatedAt}/${Asset.Name}"),
+					},
+					"s3_encryption_kms_key_arn": schema.StringAttribute{
+						CustomType: fwtypes.ARNType,
+						Optional:   true,
+						Validators: []validator.String{
+							validators.ARN(),
+						},
+					},
+					"s3_encryption_type": schema.StringAttribute{
+						Optional:   true,
+						CustomType: fwtypes.StringEnumType[awstypes.ServerSideEncryptionTypes](),
+					},
+				},
+			},
+			"event_revision_published": schema.SingleNestedBlock{
+				CustomType: fwtypes.NewObjectTypeOf[eventRevisionPublishedModel](ctx),
+				Attributes: map[string]schema.Attribute{
+					"data_set_id": schema.StringAttribute{
+						Required: true,
+					},
+				},
+			},
+		},
+	}
+}
+
+func (r *resourceEventAction) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	conn := r.Meta().DataExchangeClient(ctx)
+
+	var data resourceEventActionModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	input, diags := r.buildCreateInput(ctx, data)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
```

Comment:
Rather than custom methods to construct the input, AutoFlex should be used. At a high level, this process uses reflection to convert between the Terraform and AWS data structures. More information available here:

https://hashicorp.github.io/terraform-provider-aws/data-handling-and-conversion/#autoflex-for-terraform-plugin-framework-preferred